text/0000-glimmer-components.md

```diff
@@ -0,0 +1,1022 @@
+behind the "install your way to Ember" goal, which will enable Ember to be used
+in more constrained environments that smaller frameworks such as React, Preact,
+Vue, and more excel in. It will also allow users who are size-conscious to adopt
+Ember incrementally, adding functionality when it is _needed_ rather than having
+all-or-nothing.
+
+This will take time though. Much progress has been made, but parts of Ember are
+still very monolithic. And while it isn't Ember, Glimmer.js is a _very_
+lightweight wrapper of the Glimmer VM that enables users to drop that weight and
+begin writing much more minimal apps today.
+
+Glimmer components aren't just based on the Glimmer.js component API - they are
+one and the same. They will be a shared package, which will be importable and
+usable by the users of both frameworks. Not only will users be able to write
+better components in Ember, those components will also be cross-compatible with
+Glimmer apps (assuming they don't use Ember specific functionality).
+
+It is important to note that while Glimmer components will be versioned
+independently from Glimmer and Ember ***they will abide by the Ember RFC process
+for any and all changes to user APIs***. The implementations for their component
+managers in Glimmer and Ember may change to keep them compatible, but they will
+not make major changes without first getting community input, and will be
+considered part of the public API of Ember.
+
+## Prior Art
+
+Part of the challenge in the original Angle Bracket components RFC was
+attempting to design without implementation, testing, usage, and feedback.
+Glimmer.js provided an early method to experiment, but because it was not widely
+adopted there wasn't much feedback from larger-scale usage. This in part
+motivated the [component manager
+RFC](https://github.com/emberjs/rfcs/blob/master/text/0213-custom-components.md),
+which enabled experimentation in Ember directly, and set us up for having
+multiple implementations of component APIs which were interchangeable.
+
+As such, we now have two reference implementations which can be referred to:
+
+* [Glimmer.js](https://glimmerjs.com/), the framework that this component API is
+  based on, and will be cross-compatible with.
+* [sparkles-component](https://github.com/rwjblue/sparkles-component), an an
+  implementation of the Glimmer.js component API using Ember's component
+  managers. It is usable in Ember today.
+
+Both of these have minor differences from the API proposed in this RFC, mainly
+because they were made before the [element modifiers
+RFC](https://github.com/emberjs/rfcs/blob/master/text/0373-Element-Modifier-Managers.md)
+was accepted and opened up additional design possibilities. However, they serve
+as valuable data points for the viability of a simpler component API, and inform
+the design accordingly.
+
+## Detailed design
+
+Glimmer components have the following interface:
+
+```ts
+interface GlimmerComponentt {
```

Comment:
Seems like this should be generic over `args`?

Reply:
`GlimmerComponentt` -> `GlimmerComponent`

Reply:
> Seems like this should be generic over args?

FWIW, @mike-north and I went with a generic + default in sparkles-component.

```ts
interface Component<T = object> {
  args: T;
  constructor(owner: Owner, args: T): void;

  didInsertElement(): void;
  didUpdate(): void;
  
  willDestroy(): void;
}
```